[Quant][fx] Implement BackendConfig (part 1)

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81469

Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

andrewor14 committed:
[Quant][fx] Implement BackendConfig (part 1)

Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_config.py:
```diff
@@ -0,0 +1,323 @@
+# Owner(s): ["oncall: quantization"]
```
nit: test_config.py -> test_backend_config.py?

andrewor14 replied:
Ok. I originally wanted this test file for the other configs as well but maybe we should put them in separate test files

jerryzh168 commented on test/quantization/core/test_config.py:
```diff
@@ -0,0 +1,323 @@
+
+from torch.ao.quantization.backend_config import (
+    BackendConfig,
+    BackendOpConfig,
```
BackendPatternConfig would be more general I think

jerryzh168 commented on test/quantization/core/test_config.py:
```diff
@@ -0,0 +1,323 @@
+from torch.ao.quantization.backend_config import (
+    BackendConfig,
+    BackendOpConfig,
+    DtypeConfig,
```
nit: DTypeConfig maybe?

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -36,6 +36,7 @@
 from quantization.core.test_workflow_module import TestHistogramObserver  # noqa: F401
 from quantization.core.test_workflow_module import TestDistributed  # noqa: F401
 from quantization.core.test_workflow_module import TestFusedObsFakeQuantModule  # noqa: F401
+from quantization.core.test_config import TestBackendConfig  # noqa: F401
```
nit: test_config -> test_backend_config here as well

andrewor14 replied:
Oops I'll fix that

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+@dataclass
+class DTypeConfig:
+    """
+    Data type config for ops defined in :class:`~torch.ao.quantization.backend_config.BackendConfig`.
```
probably need more descriptions for this one, but can do this in a future PR

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+class BackendConfig:
+    # TODO: refer to native fbgemm BackendConfig once that is implemented
+    """
+    Config that defines which ops are supported and how they are quantized on a custom backend.
```
guess we need to rephrase this description now, it does not talk about ops, it's about "q/dq placement rules that are supported on this backend, the rules are described in terms of how we place q/dq ops for input and output of some patterns"

andrewor14 replied:
Ok. Should we say `quant/dequant ops` instead of `q/dq`?

jerryzh168 replied:
yeah makes sense

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+    """
+    Config that defines which ops are supported and how they are quantized on a custom backend.
+
+    This config specifies how reference quantized models are produced (during the prepare phase) and how they
```
also need to be careful about this as well, by default it only describes how a reference quantized model is produced, but for the existing operator based backends like fbgemm/qnnpack, we can also interpret the config in terms of quantized operators (how the patterns are lowered)

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+
+        linear_config = BackendPatternConfig(torch.nn.Linear) \
+            .set_observation_type(ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT) \
+            .set_dtype_configs([weighted_int8_dtype_config]) \
```
do you feel we should make this a `add_dtype_config` api instead

andrewor14 replied:
I feel it should be `set_dtype_config` to be consistent with `QConfigMapping` and `PrepareCustomConfig`. What do you think?

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+            .set_fuser_method(reverse_sequential_wrapper2(torch.nn.intrinsic.ConvReLU2d))
+
+        backend_config = BackendConfig("my_backend") \
+            .set_config(linear_config) \
```
this one should be `add_pattern_config` I think

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+        self.name = name
+        return self
+
+    def set_config(self, config: BackendPatternConfig) -> BackendConfig:
```
nit: maybe `add_backend_pattern_config`

andrewor14 replied:
(Same as above) `set_backend_pattern_config`?

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+    """
+    Config for ops defined in :class:`~torch.ao.quantization.backend_config.BackendConfig`.
+
+    The user can configure how an op is handled on a given backend using the following methods:
```
`how an op is handled` -> `how a operator pattern graph is handled`

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+        self._overwrite_output_fake_quantize: Optional[_PartialWrapper] = None
+        self._overwrite_output_observer: Optional[_PartialWrapper] = None
+
+    def set_pattern(self, pattern: Pattern) -> BackendPatternConfig:
```
maybe we don't need this method? pattern is required so maybe user can pass this as an argument for initialization?

andrewor14 replied:
Ok, makes sense

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+        self.reference_quantized_module = reference_quantized_module
+        return self
+
+    def set_fuser_module(self, fuser_module: torch.nn.Module) -> BackendPatternConfig:
```
fused_module?

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+            .set_config(linear_config) \
+            .set_config(conv_relu_config)
+    """
+    def __init__(self, name: str):
```
should name be optional as well?

andrewor14 replied:
What should the default be?

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,392 @@
+        self.root_module: Optional[torch.nn.Module] = None
+        self.qat_module: Optional[torch.nn.Module] = None
+        self.reference_quantized_module: Optional[torch.nn.Module] = None
+        self.fuser_module: Optional[torch.nn.Module] = None
```
should be `fused_module` I think

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -0,0 +1,388 @@
+
+    def add_dtype_config(self, dtype_config: DTypeConfig) -> BackendPatternConfig:
+        """
+        Set the supported data types for this op.
```
nit: comments need to be updated, also wondering if we want to give something more useful here, e.g. include some examples

jerryzh168 reviewed (approved):
looks good, thanks for adding the support and addressing all the comments, had one nit comment inline

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Implement BackendConfig (part 1)"


Summary: Following https://github.com/pytorch/pytorch/pull/78452
and https://github.com/pytorch/pytorch/pull/79066, this commit
is part 1 of the broader effort to replace `backend_config_dict`
with a python config object, a more formal and robust API that
leads to better user experience. Note that there is no change in
behavior in this commit by itself. A future commit (part 2) will
replace all existing usages of `backend_config_dict` with the
`BackendConfig` object added in this commit.

Test Plan:
python test/test_quantization.py TestBackendConfig

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge
